# Final Year Project - CNA Signatures Code Workflow
# Bioinformatic Analysis of Copy Number Alteration Signature Mechanisms in Ovarian Carcinoma
# Function code from - "Copy number signatures and mutational processes in ovarian carcinoma", Macintyre et al 2018.
# Macintyre et al. code available - https://bitbucket.org/britroc/cnsignatures/src/master/
# Conor Giles Doran - UCD - Sept. 2020 - Feb. 2021

#### NB ####

# The published CNA signature scripts contained problematic variable naming that could not be resolved.
# This was unexpected and prevented recapitulation of methods with raw CNA data.
# Recapitulation was attempted but did not work due to miscommunication of essential method components.
# The results generated by Macintrye et. al were used as input and their findings were confirmed.

#### PACKAGES ####

libs<-c("Biobase", "QDNAseq", "NMF", "flexmix", "YAPSA", "tidyr", 'dplyr')
lapply(libs, library, character.only = TRUE)


#### SETUP ####

# script directory
SCRIPTDIR <- "G:/Shared drives/BHLS/2020/Conor Giles Doran/R Scripts/Thesis_scripts/Signatures"

# signature data directory
DATADIR <- "G:/Shared drives/BHLS/data/cnsignatures/britroc-cnsignatures-bfb69cd72c50"

# output directory
OUTDIR <- "G:/Shared drives/BHLS/2020/Conor Giles Doran/Final_output/Signature_output"
dir.create(OUTDIR, showWarnings = FALSE)


#### DATA ####

# Source required functions from 'All_Sig_functions.R' script
source(paste0(SCRIPTDIR, "/All_Sig_Functions.R"))

# chromosome size and gap data
chrom_size_file <- paste0(DATADIR, "/data/hg19.chrom.sizes.txt")
chrom_gap_file <- paste0(DATADIR, "/data/gap_hg19.txt")

## NB ##

# CN feature data used here was provided by Macintyre et. al 
# If possible, it should be extracted from raw data using the extractCopynumberFeatures function 

CN_feats.rds <-  paste0(DATADIR, "/manuscript_Rmarkdown/data/tcga_CN_features.rds")

# signature definition matrix - signature_by_component matrix
# component parameters provided by Macintyre et. al

feat_sig_mat <- paste0(DATADIR, "/feat_sig_mat.rds")
feat_components.rds <- paste0(DATADIR, "/data/component_parameters.rds")

# directory for saving sample_by_comp

sample_by_comp.rds <- paste0(OUTDIR, "/sample_by_comp.rds")

#### WORKFLOW ####

#### Copy number feature extraction ####

# copy number profiles were not able to be generated from raw CN data in this project
# Therefore the CN_feats object was provided by Macintyre et. al 2018 (link above)
# In the Macintyre et. al code it can be generated by the extractCopynumberFeatures function
# whereby copy-number profiles are input as a QDNAseq object, or as a list of segment tables. 


if(!exists("CN_feats.rds")){
  print("Please provide a rds object detailing copy number feature data ...")
} else {
  CN_feats <- readRDS("CN_feats.rds")  # as provided by Macintyre et. al
}

#### MIXTURE MODELLING ####

# extract optimal number of model components for each copy number feature
# NB - model components here have been provided by Macintyre et. al

if(!exists("feat_components.rds")){
  print("Generating Feature Components...")
  feat_components <- fitMixtureModels(CN_feats)
  saveRDS(feat_components, file = "feat_components.rds")
} else {
  feat_components <- readRDS("feat_components.rds")
}

# generate the sample by component matrix using feature and component results

if(!exists("sample_by_comp.rds")){
  print("Generating Sample by Components...")
  sample_by_comp <- generateSampleByComponentMatrix(CN_feats,
                                                    all_components = feat_components)
  saveRDS(sample_by_comp, file = "sample_by_comp.rds")
} else {
  sample_by_comp <- readRDS("sample_by_comp.rds")
}

# sample_by_comp <- generateSampleByComponentMatrix(CN_feats, all_components = feat_components)

# saveRDS(sample_by_comp,
#        file = paste0(OUTDIR, "/sample_by_components.rds"))

# generate and save matrix
pdf(paste0(OUTDIR, "/TCGA_OV.sample_by_components.pdf"), onefile = FALSE)
NMF::aheatmap(sample_by_comp,
              fontsize = 14,
              cexCol = min(0.2 + 1/log10(36), 1.2),
              Rowv=FALSE,
              Colv=FALSE,
              legend = T,
              breaks=c(seq(0,199,2),500),
              main="Sample by Components")
dev.off()


#### SIGNATURE EXTRAPOLATION ####

print("Generating Signatures...")

# identify optimal number of signatures to be used
num_sigs <- chooseNumberSignatures(sample_by_comp, outfile = paste0(OUTDIR, "/TCGA_OV.num_sigs.pdf")) # PDF NOT VIEWABLE

# number of signatures to be used
NUMSIGS <- 7

# generate the signatures from sample by comp and specified number of sigs 
TCGA_signatures <- generateSignatures(sample_by_comp, NUMSIGS)

# generate patient_by_sig and sig_by_comp heatmaps and save as pdf
pdf(paste0(OUTDIR, "/initial_heatmaps.pdf"))
coefmap(TCGA_signatures,Colv="consensus",tracks=c("basis:"), main="Patient x Signature matrix")  # red/yellow maps
basismap(TCGA_signatures,Rowv=NA,main="Signature x Component matrix", cexRow = 50)
dev.off()

# extract signature by component matrix from signature results
comp_by_sig <- TCGA_signatures@fit@W

saveRDS(TCGA_signatures, file = paste0(OUTDIR, "/TCGA_signature_results.rds"))
TCGA_signatures <- readRDS(paste0(OUTDIR, "/TCGA_signature_results.rds"))
#### QUANTIFY SIGNATURES ####

print("Quantifying Signatures...")

# end result of signature extrapolation
# normalized signature by sample matrix - use signature by comp extracted from TCGA_signature
norm_sig_by_samp <- quantifySignatures(sample_by_comp, component_by_signature = comp_by_sig)

##FIX##
signature_list <- list("num_sigs" = num_sigs, "signatures" = comp_by_sig, "sig_by_samp" = norm_sig_by_samp)

saveRDS(signature_list, file = "sig_list.rds")

# saveRDS(norm_sig_by_samp, file = paste0(OUTDIR, "/normalized_sig_x_samp.rds"))

#### FORMAT SIGNATURE RESULTS ####

print("Formatting Signature Results...")

# results formatted for differential expression workflow  
# identify maximum signature per sample
# format signature results into long format and
# format suitable for merging with clinical data

DE_sig_data <- get_DE_data(norm_sig_by_samp)

saveRDS(DE_sig_data, file = paste0(OUTDIR, "/DE_sig_data.rds"))

# heatmap for normalized signature by sample matrix

pdf(paste0(OUTDIR, "/415Norm_sig_x_sample.pdf"))
heatmap(as.matrix(DE_sig_data$clinical_sigs[,-8]))
dev.off()

sig_tally_415 <- DE_sig_data$clinical_sigs %>%
  group_by(max_signature) %>%    
  tally()

saveRDS(sig_tally_415, file = paste0(OUTDIR, "/sig_tally_415.rds"))

#### session info ####
writeLines(capture.output(sessionInfo()), paste0(OUTDIR, "/sessionInfo.txt"))
